rust/operator-binary/src/controller.rs

```diff
@@ -1444,8 +1455,17 @@ async fn get_proxy_hosts(
     Ok(proxy_hosts.join(","))
 }
 
-pub fn error_policy(_obj: Arc<NifiCluster>, _error: &Error, _ctx: Arc<Ctx>) -> Action {
-    Action::requeue(*Duration::from_secs(10))
+pub fn error_policy(
+    _obj: Arc<DeserializeGuard<NifiCluster>>,
+    error: &Error,
+    _ctx: Arc<Ctx>,
+) -> Action {
+    match error {
+        // root object is invalid, will be requeued when modified anyway
```

Comment:
I think a test would be good, to illustrate what is covered/not covered by this change.

For example, I tested this by temporarily changing an integration test to deploy an extra dud nifi-cluster definition that references a non-existent `nifi-sensitive-property-key`, and it errors out once with:

`2024-10-07T11:31:52.720550Z ERROR stackable_operator::logging::controller: Failed to reconcile object controller.name="nificluster.nifi.stackable.tech" error=reconciler for object NifiCluster.v1alpha1.nifi.stackable.tech/pre-test-nifi.kuttl-test-proven-clam failed error.sources=[security failure, sensistive key failure, sensitive key secret [nifi-sensitive-property-key/kuttl-test-proven-clam] is missing, but auto generation is disabled]`

followed by regular

`2024-10-07T11:32:02.804929Z  INFO reconciling object{object.ref=NifiCluster.v1alpha1.nifi.stackable.tech/pre-test-nifi.kuttl-test-proven-clam object.reason=related object updated: Service.v1./pre-test-nifi.kuttl-test-proven-clam}: stackable_nifi_operator::controller: Checking for sensitive key configuration`

and the test completes.

If I change other things, though (like omitting a mandatory field such as `zookeeperConfigMapName`), then the deployment (and test) fails and the behaviour is different.

Reply:
In reading the parent issue again, I don't think I'm even testing the correct issue (deserialization). So a test would definitely help!

Reply:
There are a few issues that prevent automatic testing here:

1. Triggering it (in a detectable way) is kind of racy and timing-sensitive. It would also end up interfering with unrelated tests in undesirable ways.
2. This is a guardrail against issues that "should" be impossible to trigger normally, since they would be caught by K8s' CRD validation. The one case where it wouldn't happen is during an SDP upgrade, since some objects would be validated against the old version.

In particular, neither of your cases actually trigger this code path:

> For example, I tested this by temporarily changing an integration test to deploy an extra dud nifi-cluster definition that references a non-existent nifi-sensitive-property-key, and it errors out once with:

This object has a valid shape, it "just" has garbage data that makes reconciliation impossible. This is already covered by our existing error handling.

> If I change other things, though (like omitting a mandatory field such as zookeeperConfigMapName), then the deployment (and test) fails and the behaviour is different.

Yes, that's failing to deploy the object altogether, since Kubernetes rejects it during validation.

---

To actually trigger the issue, you need to use a `kube::Resource` that is more restrictive than the corresponding Kubernetes resource, such as [the upstream example](https://github.com/kube-rs/kube/blob/main/examples/errorbounded_configmap_watcher.rs).

Reply:
OK, understood. Thank you for the clarification! (maybe testing is something we could do when we have CRD versioning).